File: dart/fourdgs/lib/src/decoded_state_budget.dart
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// Aggregate decoded-state accounting for collecting reader operations.
///
/// Per-record ceilings keep one Chunk bounded. A collecting call also needs a
/// ceiling over the state it has retained plus the next decode or composition
/// working set; otherwise many individually legal records can still exhaust a
/// process.
library;

import 'dart:typed_data';

import 'chunk_decoder.dart';
import 'exceptions.dart';

/// The shared default for collecting decoded gaussian state: 512 MiB.
const int defaultMaxDecodedStateBytes = 536870912;

/// Validates the public decoded-state budget option.
///
/// The static [int] type rejects booleans, fractional values and non-finite
/// numbers at the call boundary. This check owns the remaining invalid values.
int validateMaxDecodedStateBytes(int value) {
  if (value <= 0) {
    throw ArgumentError.value(
      value,
      'maxDecodedStateBytes',
      'must be a positive integer',
    );
  }
  return value;
}

/// Tracks decoded bytes retained by one collecting operation.
///
/// This is public so a collector assembled from incremental APIs can share the
/// same accounting with those APIs. Incremental callers normally omit it: once
/// a yielded value belongs to the caller, it no longer accrues to the library's
/// working set.
class FourdgsDecodedStateBudget {
  FourdgsDecodedStateBudget([
    int maxDecodedStateBytes = defaultMaxDecodedStateBytes,
  ]) : limit = validateMaxDecodedStateBytes(maxDecodedStateBytes);

  /// The caller-selected decoded-state ceiling.
  final int limit;

  int _retainedBytes = 0;

  /// Decoded state already retained by the collecting result.
  int get retainedBytes => _retainedBytes;

  /// Bytes still available before the configured ceiling is reached.
  int get remainingBytes => limit - _retainedBytes;

  Never _exceeded(String phase, String required) {
    throw FourdgsReaderLimit(
      'decoded-state resource limit during $phase: $required; configured '
      'limit is $limit bytes',
    );
  }

  /// Checks a simultaneous decoded allocation without retaining it.
  void check(int additionalBytes, String phase) {
    if (additionalBytes < 0) {
      throw StateError(
        'decoded-state accounting cannot subtract through check',
      );
    }
    checkBigInt(BigInt.from(additionalBytes), phase);
  }

  /// Checks an exact size expression that may exceed the platform [int] range.
  void checkBigInt(BigInt additionalBytes, String phase) {
    if (additionalBytes.isNegative) {
      throw StateError(
        'decoded-state accounting cannot subtract through check',
      );
    }
    final required = BigInt.from(_retainedBytes) + additionalBytes;
    if (required > BigInt.from(limit)) {
      _exceeded(phase, 'at least $required bytes are required');
    }
  }

  /// Checks `rows * bytesPerRow` without allowing size arithmetic to wrap.
  void checkRows(int rows, int bytesPerRow, String phase) {
    if (rows < 0 || bytesPerRow <= 0) {
      throw StateError(
        'decoded-state row accounting requires non-negative rows and a '
        'positive row width',
      );
    }
    if (rows > remainingBytes ~/ bytesPerRow) {
      final required =
          BigInt.from(_retainedBytes) +
          BigInt.from(rows) * BigInt.from(bytesPerRow);
      _exceeded(phase, 'at least $required bytes are required');
    }
  }

  /// Charges decoded output that remains owned by the collecting result.
  void retain(int addedBytes, String phase) {
    check(addedBytes, phase);
    _retainedBytes += addedBytes;
  }

  /// Releases a buffer that a collector replaced rather than returned.
  void release(int releasedBytes) {
    if (releasedBytes < 0 || releasedBytes > _retainedBytes) {
      throw StateError('decoded-state accounting released non-retained bytes');
    }
    _retainedBytes -= releasedBytes;
  }
}

/// Element capacity retained by one decoded gaussian-birth Chunk.
int decodedChunkStateBytes(FourdgsDecodedChunk chunk) {
  int bytes =
      chunk.positions.lengthInBytes +
      chunk.scales.lengthInBytes +
      chunk.rotations.lengthInBytes +
      chunk.colors.lengthInBytes +
      chunk.motions.lengthInBytes +
      chunk.muT.lengthInBytes +
      chunk.sigmaT.lengthInBytes +
      chunk.winLo.lengthInBytes +
      chunk.winHi.lengthInBytes +
      chunk.windowIndex.lengthInBytes;
  bytes += chunk.sourceGroup?.lengthInBytes ?? 0;
  bytes += chunk.sourceIndex?.lengthInBytes ?? 0;
  bytes += chunk.objectId?.lengthInBytes ?? 0;
  for (final band in chunk.shBands.values) {
    bytes += band.lengthInBytes;
  }
  return bytes;
}

/// Element capacity the final gaussian-birth result is about to allocate.
///
/// The merged SH buffer is passed through rather than copied and is therefore
/// charged by the caller at the merge step, not here.
int gaussianSetAssemblyBytes(List<FourdgsDecodedChunk> chunks) {
  int count = 0;
  for (final chunk in chunks) {
    count += chunk.count;
  }
  int bytesPerGaussian = 21 * 4;
  if (chunks.any((chunk) => chunk.sourceGroup != null)) {
    bytesPerGaussian += 4;
  }
  if (chunks.any((chunk) => chunk.sourceIndex != null)) {
    bytesPerGaussian += 4;
  }
  if (chunks.any((chunk) => chunk.objectId != null)) {
    bytesPerGaussian += 4;
  }
  return count * bytesPerGaussian;
}

/// Capacity of the single SH buffer produced by [mergeChunkBands].
int mergedShStateBytes(List<int> counts, List<Map<int, Uint8List>> chunkBands) {
  final present = <int>{for (final bands in chunkBands) ...bands.keys};
  if (present.isEmpty) return 0;
  final highest = present.reduce((a, b) => a > b ? a : b);
  final coefficients = shBandRange[highest]?.last ?? 0;
  final total = counts.fold<int>(0, (sum, count) => sum + count);
  return total * 3 * coefficients;
}

File: typescript/core/src/decodedStateBudget.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/** Aggregate decoded-state accounting for collecting reader operations. */

import type { ChunkGaussians } from "./chunk.js";
import { ExceedsReaderLimit } from "./errors.js";

/** The shared default for collecting decoded gaussian state: 512 MiB. */
export const DEFAULT_MAX_DECODED_STATE_BYTES = 536_870_912;

/** Validate the public positive-integer decoded-state ceiling. */
export function validateMaxDecodedStateBytes(value: number): number {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new RangeError("maxDecodedStateBytes must be a positive safe integer");
  }
  return value;
}

/** One collecting call's retained decoded-state bytes and peak checks. */
export class DecodedStateBudget {
  readonly limit: number;
  #retainedBytes = 0;

  constructor(limit = DEFAULT_MAX_DECODED_STATE_BYTES) {
    this.limit = validateMaxDecodedStateBytes(limit);
  }

  /** Decoded state already owned by the collecting result. */
  get retainedBytes(): number {
    return this.#retainedBytes;
  }

  /** Bytes still available before the configured ceiling is reached. */
  get remainingBytes(): number {
    return this.limit - this.#retainedBytes;
  }

  /** Check a simultaneous allocation without retaining it. */
  check(additionalBytes: number | bigint, phase: string): number {
    let additional: bigint;
    if (typeof additionalBytes === "number") {
      if (!Number.isSafeInteger(additionalBytes) || additionalBytes < 0) {
        throw new Error("decoded-state accounting requires a non-negative safe integer");
      }
      additional = BigInt(additionalBytes);
    } else {
      if (additionalBytes < 0n) {
        throw new Error("decoded-state accounting cannot subtract through check");
      }
      additional = additionalBytes;
    }
    const required = BigInt(this.#retainedBytes) + additional;
    if (required > BigInt(this.limit)) {
      throw new ExceedsReaderLimit(
        `decoded-state resource limit during ${phase}: at least ${required} bytes are required; ` +
          `configured limit is ${this.limit} bytes`,
      );
    }
    return Number(required);
  }

  /** Check `rows * bytesPerRow` using arithmetic that cannot wrap. */
  checkRows(rows: number, bytesPerRow: number, phase: string): number {
    if (
      !Number.isSafeInteger(rows) ||
      rows < 0 ||
      !Number.isSafeInteger(bytesPerRow) ||
      bytesPerRow <= 0
    ) {
      throw new Error(
        "decoded-state row accounting requires non-negative safe-integer rows and a positive width",
      );
    }
    return this.check(BigInt(rows) * BigInt(bytesPerRow), phase);
  }

  /** Charge decoded output that remains owned by the collecting result. */
  retain(addedBytes: number, phase: string): number {
    const retained = this.check(addedBytes, phase);
    this.#retainedBytes = retained;
    return retained;
  }

  /** Release a replaced buffer that the collector no longer owns. */
  release(releasedBytes: number): void {
    if (
      !Number.isSafeInteger(releasedBytes) ||
      releasedBytes < 0 ||
      releasedBytes > this.#retainedBytes
    ) {
      throw new Error("decoded-state accounting released non-retained bytes");
    }
    this.#retainedBytes -= releasedBytes;
  }
}

/** Element capacity retained by one decoded gaussian-birth Chunk. */
export function decodedChunkStateBytes(chunk: ChunkGaussians): number {
  return (
    chunk.positions.byteLength +
    chunk.scales.byteLength +
    chunk.rotations.byteLength +
    chunk.colors.byteLength +
    chunk.motions.byteLength +
    chunk.muT.byteLength +
    chunk.sigmaT.byteLength +
    chunk.windowIndex.byteLength +
    (chunk.sourceGroup?.byteLength ?? 0) +
    (chunk.sourceIndex?.byteLength ?? 0) +
    (chunk.objectId?.byteLength ?? 0)
  );
}

/** Capacity of the arrays allocated by final gaussian-birth assembly. */
export function gaussianSetAssemblyBytes(chunks: readonly ChunkGaussians[]): bigint {
  let count = 0n;
  let sourceGroup = false;
  let sourceIndex = false;
  let objectId = false;
  for (const chunk of chunks) {
    count += BigInt(chunk.count);
    sourceGroup ||= chunk.sourceGroup !== null;
    sourceIndex ||= chunk.sourceIndex !== null;
    objectId ||= chunk.objectId !== null;
  }
  // 21 f32 lanes: position, scale, rotation, colour, motion, mu, sigma, winLo, winHi.
  const identityLanes = Number(sourceGroup) + Number(sourceIndex) + Number(objectId);
  return count * BigInt(21 * Float32Array.BYTES_PER_ELEMENT + identityLanes * 4);
}
